File: frontend/desktop/src/store/modules/project.js
/**
* Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
* Edition) available.
* Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
* Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* http://opensource.org/licenses/MIT
* Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
* an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
* specific language governing permissions and limitations under the License.
*/
import axios from 'axios'

const project = {
    namespaced: true,
    state: {
        project_id: window.DEFAULT_PROJECT_ID,
        bizId: '',
        projectName: '',
        userProjectList: [], // 用户有权限的项目列表
        timeZone: window.TIMEZONE,
        authActions: []
    },
    mutations: {
        setUserProjectList (state, data) {
            state.userProjectList = data
        },
        setProjectId (state, id) {
            let projectId = id
            if (typeof id !== 'number') {
                projectId = isNaN(Number(id)) || id === '' ? '' : Number(id)
            }
            state.project_id = projectId
        },
        setBizId (state, id) {
            state.bizId = id
        },
        setTimeZone (state, data) {
            state.timeZone = data
        },
        setProjectName (state, data) {
            state.projectName = data
        },
        setProjectActions (state, data) {
            state.authActions = data
        }
    },
    actions: {
        // 更改用户的默认项目
        changeDefaultProject ({ state }, id) {
            return axios.post(`core/api/change_default_project/${id}/`).then(response => response.data)
        },
        // 加载用户有权限的项目列表
        loadUserProjectList ({ commit }, params) {
            return axios.get(`api/v3/user_project/`, { params }).then((response) => {
                if (params.limit === 0) { // 拉全量项目时更新项目列表，区分项目管理页面的分页数据
                    commit('setUserProjectList', response.data.objects)
                }
                return response.data
            })
        },
        // 获取常用业务
        loadCommonProject ({ commit }, data) {
            return axios.get('api/v3/common_use_project/').then(response => response.data)
        },
        // 获取环境变量列表
        loadEnvVariableList ({ commit }, params) {
            return axios.get(`api/v3/project_constants/`, { params }).then(response => response.data)
        },
        // 新增环境变量
        createEnvVariable ({ commit }, data) {
            return axios.post(`api/v3/project_constants/`, data).then(response => response.data)
        },
        // 更新环境变量
        updateEnvVariable ({ commit }, data) {
            const { id } = data
            return axios.put(`api/v3/project_constants/${id}`, data).then(response => response.data)
        },
        // 删除环境变量
        deleteEnvVariable ({ commit }, id) {
            return axios.delete(`api/v3/project_constants/${id}/`).then(response => response.data)
        },

        createProject ({ commit }, data) {
            const { name, time_zone, desc } = data

            return axios.post(`api/v3/project/`, {
                name,
                time_zone,
                desc
            }).then(response => response.data)
        },
        loadProjectDetail ({ commit }, id) {
            return axios.get(`api/v3/project/${id}/`).then(response => response.data)
        },
        // 更新项目详情
        updateProject ({ commit }, data) {
            const { id, name, time_zone, desc, is_disable } = data
            return axios.patch(`api/v3/project/${id}/`, {
                name,
                time_zone,
                desc,
                is_disable
            }).then(response => response.data)
        },
        getProjectConfig ({ commit }, id) {
            return axios.get(`api/v3/project_config/${id}/`).then(response => response.data)
        },
        updateProjectConfig ({ commit }, params) {
            const { id, executor_proxy, executor_proxy_exempts } = params
            return axios.patch(`api/v3/project_config/${id}/`, { executor_proxy, executor_proxy_exempts }).then(response => response.data)
        },
        getProjectStaffGroupList ({ commit }, params) {
            return axios.get(`api/v3/staff_group/`, { params }).then(response => response.data)
        },
        createProjectStaffGroup ({ commit }, params) {
            return axios.post(`api/v3/staff_group/`, params.data).then(response => response.data)
        },
        updateProjectStaffGroup ({ commit }, params) {
            const { id, data } = params
            return axios.put(`api/v3/staff_group/${id}/`, data).then(response => response.data)
        },
        delProjectStaffGroup ({ commit }, id) {
            return axios.delete(`api/v3/staff_group/${id}/`).then(response => response.data)
        },
        // 查询项目下用户可编辑的模板标签
        getProjectLabels ({ commit }, id) {
            return axios.get('api/v3/new_label/', {
                params: { project_id: id }
            }).then(response => response.data)
        },
        // 查询项目下支持的模板标签（包含默认标签）
        getProjectLabelsWithDefault ({ commit }, id) {
            return axios.get('api/v3/new_label/list_with_default_labels/', {
                params: { project_id: id }
            }).then(response => response.data)
        },
        createTemplateLabel ({ commit }, data) {
            return axios.post(`api/v3/new_label/`, data).then(response => response.data)
        },
        updateTemplateLabel ({ commit }, data) {
            return axios.put(`api/v3/new_label/${data.id}/`, data).then(response => response.data)
        },
        delTemplateLabel ({ commit }, id) {
            return axios.delete(`api/v3/new_label/${id}/`).then(response => response.data)
        },
        getlabelsCitedCount ({ commit }, payload) {
            const { ids, project_id } = payload
            return axios.get('api/v3/new_label/get_label_template_ids/', {
                params: { label_ids: ids, project_id }
            }).then(response => response.data)
        }
    }
}

export default project

File: gcloud/iam_auth/view_interceptors/project.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

from iam import Action, Subject
from iam.exceptions import AuthFailedException
from iam.shortcuts import allow_or_raise_auth_failed

from gcloud.core.models import Project
from gcloud.iam_auth import IAMMeta, get_iam_client, res_factory
from gcloud.iam_auth.intercept import ViewInterceptor

iam = get_iam_client()


class _ProjectActionInterceptor(ViewInterceptor):
    """Django function view 通用项目级 IAM 校验拦截器基类。

    APIGW 侧的 ProjectViewInterceptor 依赖 ``@project_inject`` 注入的
    ``request.project``，而 Django 侧（taskflow3/tasktmpl3/contrib 等）一般
    通过 URL 路径以 ``kwargs["project_id"]`` 传入项目 ID，因此提供一份独立的
    基于 kwargs 的实现，避免在每个视图模块重复书写鉴权样板。
    """

    action = None

    def process(self, request, *args, **kwargs):
        project_id = kwargs.get("project_id")
        if project_id is None:
            project_id = request.GET.get("project_id") or request.POST.get("project_id")

        subject = Subject("user", request.user.username)
        action = Action(self.action)

        # project_id 缺失或指向不存在的项目时，统一按鉴权失败(403)处理：
        # 1. 避免 resources_for_project 内部 Project.objects.get(id=None/不存在) 抛
        #    DoesNotExist 造成 500，攻击者可借 500/403 差异探测拦截器是否挂载；
        # 2. 缺少项目作用域的请求本就无法通过对象级鉴权，直接拒绝更安全。
        if not project_id:
            raise AuthFailedException(IAMMeta.SYSTEM_ID, subject, action, [])
        try:
            resources = res_factory.resources_for_project(project_id)
        except Project.DoesNotExist:
            raise AuthFailedException(IAMMeta.SYSTEM_ID, subject, action, [])

        allow_or_raise_auth_failed(
            iam=iam,
            system=IAMMeta.SYSTEM_ID,
            subject=subject,
            action=action,
            resources=resources,
            cache=True,
        )


class ProjectViewInterceptor(_ProjectActionInterceptor):
    """项目查看级权限校验，对应 IAMMeta.PROJECT_VIEW_ACTION。"""

    action = IAMMeta.PROJECT_VIEW_ACTION


class ProjectFlowCreateInterceptor(_ProjectActionInterceptor):
    """项目下流程创建权限校验，对应 IAMMeta.FLOW_CREATE_ACTION。

    用于 import / check_before_import 等"导入前置探测"接口，与正式导入接口的
    项目级权限保持一致，避免攻击者借助预检接口探测目标项目的流程冲突情况。
    """

    action = IAMMeta.FLOW_CREATE_ACTION
